🎉 New Source: Firebase Realtime Database

koji-m opened the pull request:
## What
* Add new source Firebase Realtime Database
* Participating in [hacktoberfest-2022](https://github.com/airbytehq/connector-contest)
* This PR closes https://github.com/airbytehq/connector-contest/issues/20

## How
Supports single stream, beacause Firebase Realtime Database is a schema-less database.
[The database's data is a JSON tree.](https://firebase.google.com/docs/database/rest/structure-data#how_data_is_structured_its_a_json_tree)

The stream to sync is a JSON object, specified by node path (ex. `path/to/users` ) in config,  which is subtree of the JSON tree.

## Recommended reading order
1. README.md
1. spec.json
1. source.py
1. firebase_rtdb.py

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</details>

<details><summary><strong>Connector Generator</strong></summary>

- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed

</details>

## Tests

<details><summary><strong>Unit</strong></summary>

*Put your unit tests output here.*

</details>

<details><summary><strong>Integration</strong></summary>

*Put your integration tests output here.*

</details>

<details><summary><strong>Acceptance</strong></summary>

*Put your acceptance tests output here.*

</details>

koji-m commented:
`./gradlew` :airbyte-integrations:connectors:source-firebase-realtime-database:integrationTest` fails even though I set the dependencies (firebase-admin) in setup.py.

```
ImportError: Error importing plugin "integration_tests.acceptance": No module named 'firebase_admin'
```

sajarin commented:
Hey @koji-m, 

Happy Hacktoberfest! Thanks for the PR and the new connector. This PR will be reviewed by a member of the community as part of Airbyte's Maintainer program. @tuanchris has been assigned to review this PR and help it get merged. Thanks again!

tuanchris commented:
Hi @koji-m, thank you for the PR. Can you share screenshots/logs of the integration tests passing?

tuanchris commented:
> Hi @tuanchris , I've uploaded screeshots of the integration test results.
> 
> The test by pytest passes in the venv environment, but the test by gradle executed from the repository root does not pass for some reason.
> 
> File path referenced in Docker container created by `integration_tests/acceptance.py` may be the problem.
> 
> ```
> "failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: exec: "/integration_tests/setup_data.sh": stat /integration_tests/setup_data.sh: no such file or directory: unknown"
> ```
> 
> integration test failure logs

Did you build the connector? Be sure to build it first using the following command at root: 

```
./gradlew :airbyte-integrations:connectors:source-firebase-realtime-database:build
```

You can follow instructions [here](https://github.com/airbytehq/airbyte/blob/be3dcfe0f551f995ffcdf06bad55a0710129a9d9/airbyte-integrations/connectors/source-firebase-realtime-database/README.md) if you haven't done so already. Let me know if it helpes.

koji-m commented:
> Did you build the connector?

Yes, I did. 

I think the test container's (ran for setting up test data from acceptance.py) volume mount path is different depending on execution commands (python or gradlew).

koji-m commented:
I could build the connecto locally.
![image](https://user-images.githubusercontent.com/7721142/196972998-322283bc-a803-4341-9911-32840baf8282.png)

koji-m commented:
@tuanchris It looks failing to pip install grpcio. Could you try the following command to see if it works?

```shell
docker run -it --rm  python:3.9.11-alpine3.15 pip install firebase-admin
```

koji-m commented:
@tuanchris The acceptance test now passes. I've put the screen shot of the test results in the description.

tuanchris commented on airbyte-integrations/connectors/source-firebase-realtime-database/integration_tests/abnormal_state.json:
```diff
@@ -0,0 +1,5 @@
+{
```
You can remove this file since we don't have an incremental stream

koji-m replied:
fixed in https://github.com/airbytehq/airbyte/pull/18029/commits/ea934df62a530ad26eceaeeb4c7a344e49938806

tuanchris commented on airbyte-integrations/connectors/source-firebase-realtime-database/integration_tests/catalog.json:
```diff
@@ -0,0 +1,39 @@
+{
+  "streams": [
+    {
+      "name": "TODO fix this file",
+      "supported_sync_modes": ["full_refresh", "incremental"],
+      "source_defined_cursor": true,
+      "default_cursor_field": "column1",
```
Can you fill in the values in this file?

koji-m replied:
fixed in https://github.com/airbytehq/airbyte/pull/18029/commits/ae71001b07290a3d91b24d2fc195e5b096d76020

tuanchris reviewed (changes requested):
Hi @koji-m, I was able to build locally and successfully test the connector. I left some comments below. Please have a look.

Was you able to run these successfully? I was not able to do so locally. 
```python
docker run --rm -v $(pwd)/secrets:/secrets airbyte/source-firebase-realtime-database:dev check --config /secrets/config.json
docker run --rm -v $(pwd)/secrets:/secrets airbyte/source-firebase-realtime-database:dev discover --config /secrets/config.json
docker run --rm -v $(pwd)/secrets:/secrets -v $(pwd)/integration_tests:/integration_tests airbyte/source-firebase-realtime-database:dev read --config /secrets/config.json --catalog /integration_tests/configured_catalog.json
```

koji-m commented on airbyte-integrations/connectors/source-firebase-realtime-database/source_firebase_realtime_database/spec.yaml:
```diff
@@ -0,0 +1,27 @@
+    database_name:
+      title: Database Name
+      type: string
+      description: Database name
```
fixed in https://github.com/airbytehq/airbyte/pull/18029/commits/07cab9bfcbb04363be20d388a102949ab15c11a8

koji-m commented on airbyte-integrations/connectors/source-firebase-realtime-database/source_firebase_realtime_database/spec.yaml:
```diff
@@ -0,0 +1,27 @@
+    google_application_credentials:
+      title: Service Account Key JSON
+      type: string
+      description: JSON cert credentials in JSON format
```
fixed in https://github.com/airbytehq/airbyte/pull/18029/commits/043b2dcc78c6c70783fd42b08b84696d8d173712

koji-m commented:
@tuanchris 

Thank you for your review. I fixed the codes you pointed out.

> Was you able to run these successfully? I was not able to do so locally.

`supported_sync_modes` property in `integration_tests/configured_catalog.json` was missing. I fixed in https://github.com/airbytehq/airbyte/pull/18029/commits/733152b6dc6c1e9484dbe7404ec28965859b78ac and was able to docker run successfully as shown below.

![image](https://user-images.githubusercontent.com/7721142/197329606-6aad817e-a9ed-45e0-b75c-6bee63556aa0.png)

I set the test data as follows before run.

![image](https://user-images.githubusercontent.com/7721142/197329554-6c7f44e2-a2cb-4edd-a057-414a7f774e8e.png)

tuanchris reviewed (approved):
@koji-m Thank you. Everything looks good now 
@sajarin FYI.

tuanchris commented:
@koji-m none from my end. @sajarin can you take a look at finalizing this PR?

sajarin commented:
/publish connector=connectors/source-firebase-realtime-database run-tests=false

sajarin commented:
/publish connector=connectors/source-firebase-realtime-database run-tests=false

sajarin commented:
/publish connector=connectors/source-firebase-realtime-database run-tests=false

koji-m commented:
@sajarin Is there any problem?

marcosmarxm commented:
/publish connector=connectors/source-firebase-realtime-database run-tests=false

marcosmarxm commented:
Sorry the delay to get this merge @koji-m I'll solve this tomorrow.

marcosmarxm commented:
Looks there is a problem installing the libs.

koji-m commented:
I couldn't reproduce the build error locally. 
```
#9 2796.4   Building wheel for cffi (setup.py): finished with status 'error'
...
5535.0   Running setup.py install for grpcio: finished with status 'error'
```

@marcosmarxm Can you reproduce the error locally?

marcosmarxm commented:
Hello :wave:, first thank you for this amazing contribution. 

We really appreciate the effort you've made to improve the project.
We ask you patience for the code review. Last month our team was focused on Hacktoberfest event and that probably left some PR without the proper feedback. And this week, due to the Thanksgiving US Holiday, most our team is out of office with their families. Another important piece of information why code won't be merge this week is: as a safety measure the core team has decided to freeze merging code to main branch to keep the release stable. Next week we'll return to you with the proper code review and update the status of your contribution.

If you have any questions feel free to send me a message in Slack!
Thanks!

marcosmarxm commented:
/publish connector=connectors/source-firebase-realtime-database run-tests=false

marcosmarxm commented:
/publish connector=connectors/source-firebase-realtime-database run-tests=false

marcosmarxm commented:
@koji-m sorry the delay here. The problem with `grpcio` continues. I'd requested help to the connector team.

marcosmarxm commented:
/publish connector=connectors/source-firebase-realtime-database run-tests=false

marcosmarxm commented:
/publish connector=connectors/source-firebase-realtime-database run-tests=false

marcosmarxm commented:
Hello :wave::skin-tone-2: and thank you for your contribution!

Airbyte has instituted a code freeze between 19 and 30 December, to make sure there are no disruptions during the holidays. 
Because of this, reviewing and merging your contribution may take longer than usual. 
We apologize for the delay, but we want everyone to have a quiet and happy holiday.

If you have any questions or need further clarification, please don't hesitate to ping via Slack.

TBernstein4 commented:
@marcosmarxm Can you take a look at this PR when you return and see where it stands?
